### ISSUE TITLE
Storage emulator feature request

### ISSUE THREAD

**Issue by nicobrinkkemper** [firebase/firebase-tools#1738]
It's cool and all when github isn't the place for feature requests. But what if the feature request page isn't working?

![image](https://user-images.githubusercontent.com/1597349/67159104-849f4e00-f340-11e9-9de7-e0a896183c0f.png)

We need a storage emulator so we can start testing our AutoML features locally. Are there any workarounds? Hacks? Thanks in advance,

Regards,

Jim

**Comment by bkendall** [comment]
PIng @samtstern as he has done a lot with the emulators.

Sorry to hear about the request page not working (and sorry for the super late response). It probably was a transient issue if you'd like to try again...

**Comment by samtstern** [comment]
Thanks for this!  We are also interested in a Storage emulator, but no timeline to share right now.

**Comment by otri** [comment]
Hoping for some storage emulator love.  ❤️ 🤖

**Comment by davigp** [comment]
I'd like to debug and test a storage function `onFinalize`, but seems it still not possible:

> `functions[generateThumbnail]: function ignored because the storage emulator does not exist or is not running.`

How can I debug a `onFinalize` function locally without the storage emulator? 
Is there any way to create unit tests for storage functions?

**Comment by samtstern** [comment]
@davigp you can use `functions:shell` to test those functions interactively or you can use the `functions-testing` library to write unit tests.

**Comment by sebolio** [comment]
Any news on this? Sounds really cool

**Comment by samhorlbeck** [comment]
Hi folks, thanks for the interest! We're considering this, internal tracking bug b/159057998

**Comment by vanities** [comment]
> functions-testing

can you link the repo? Do you mean the `firebase-functions-test` library?

**Comment by samtstern** [comment]
@vanities yes sorry that's what I meant, there are some examples here:
https://github.com/firebase/quickstart-testing

**Comment by vanities** [comment]
@samtstern 
No worries, thanks again! Hoping we can get emulator testing for storage and functions, the emulator is so useful!

**Comment by IronDigitalMedia** [comment]
We are also interested in a storage emulator! Why wasn't this thought of? If so, if they couldn't do it why not just get ahead of it and post a blog post explaining the difficulties?

**Comment by tzoratto** [comment]
The Firestore emulator lets us test our security rules in a automated way on every build (https://firebase.google.com/docs/firestore/security/test-rules-emulator?hl=en) and on developer computers in one click, without any credentials or anything

Having the same workflow/feature for storage security rules would be awesome.

Automated and easy security testing is a game changer for me, I hope it will expand to storage as well 👍

**Comment by rdinkel** [comment]
We are also in need of the same awesome emulator testing feature for storage rules. Please push this request in its priority :100:  ;)

**Comment by warting** [comment]
+1 this would be awesome!

**Comment by HerrNiklasRaab** [comment]
@samtstern After authentication emulator, this one will be the next? :)

**Comment by valentinoavon** [comment]
I'd love a storage emulator! :D

**Comment by mhswork** [comment]
Firebase Storage is like the heart of firebase when we have projects dealing with storing images, videos etc. Once we get this in an Emulator then that would be a huge relief. Waiting for its release.

**Comment by ribomation** [comment]
It's awkward to handle two different credentials, one for the emulator suite and another one when accessing firebase storage. 
What is the time frame for making the storage emulator available? A month from now, before end of year or first next year?

**Comment by samtstern** [comment]
@ribomation sorry we can't offer any timelines but I can say definitely longer than a month!

**Comment by chrisjacobs91** [comment]
Hi all - I would also love this emulator. In the meantime, is there a way I can avoid deleting live storage files and folders when running hosting / functions in the emulator? 

For example I use the following code to delete all the files in a folder. Last night someone accidentally deleted all the documents in our emulator as part of a test and it deleted all the LIVE storage folders as we use an import of real documents into our emulator 🤦 

```
async function deleteStorageFolder(path:string) {
  const bucket = admin.storage().bucket();
  return bucket.deleteFiles({
    prefix: path
  })
```

Is there any way I can tell firebase to avoid using the production storage APIs when emulators are running?

**Comment by itsravenous** [comment]
> Is there any way I can tell firebase to avoid using the production storage APIs when emulators are running?

I use separate Firebase projects for dev/staging/ci. You don't have to get that granular, but having a non-prod project would help you avoid that.

But of course, really, I just want a storage emulator just like everyone else.

**Comment by Develrockment** [comment]
The storage emulator would be very helpful indeed

**Comment by premsuman8** [comment]
We need the storage emulator!
Give the people what they want! :trollface:

**Comment by qbolt** [comment]
Just started using firebase and loving it so far. Only big pain-point for me so far is no storage emulator. :( 

Hoping to see it soon!

**Comment by ChristianTucker** [comment]
Mindless +1.

**Comment by maranmaran** [comment]
+1

**Comment by jsduffy** [comment]
+1

**Comment by isamyh** [comment]
@samtstern 

Any news? <3

**Comment by eladdad** [comment]
Is there a way to use Firebase Auth emulator with storage until there is a storage emulator? I need to emulate users uploading directly to storage with security rules applied

**Comment by bofeiw** [comment]
Storage emulator would be very useful to our automated testing. Hope it'll be available soon.

**Comment by bramosrys** [comment]
+1 Cannot test with security rules enabled because file access is denied due the Firebase Auth emulator running :(

**Comment by rip3rs** [comment]
Hi all, I'm here for the storage emulator :)

btw:
> > Is there any way I can tell firebase to avoid using the production storage APIs when emulators are running?
> 
> @chrisjacobs91 I use separate Firebase projects for dev/staging/ci. You don't have to get that granular, but having a non-prod project would help you avoid that.
> 
> But of course, really, I just want a storage emulator just like everyone else.

@itsravenous that can get expensive. Isn't that a bit dangerous?

**Comment by itsravenous** [comment]
> @itsravenous that can get expensive. Isn't that a bit dangerous? having multiple envs on firebase?

That's of course your call. I haven't scratched the surface of my free limit via automated tests run against my non-prod firebase env, but obviously that entirely depends on how many tests you have, how often you run them, and what they are doing :smile: 

I would certainly advise you set up billing alerts to ensure you don't spend more than you want to.

**Comment by bofeiw** [comment]
> > @itsravenous that can get expensive. Isn't that a bit dangerous? having multiple envs on firebase?
> 
> That's of course your call. I haven't scratched the surface of my free limit via automated tests run against my non-prod firebase env, but obviously that entirely depends on how many tests you have, how often you run them, and what they are doing 😄
> 
> I would certainly advise you set up billing alerts to ensure you don't spend more than you want to.

This could be a temporary workaround, not ideal for larger scale automated testing. Running emulator locally will be much time-efficient to run the test cases.

**Comment by itsravenous** [comment]
> This could be a temporary workaround, not ideal for larger scale automated testing. Running emulator locally will be much time-efficient to run the test cases.

Totally agree. For me, it works OK as my project is fairly small. But of course, if we could emulate storage then that would be much, much better.

**Comment by Canavell** [comment]
We need this! It's even looks strange when you have emulators on all your products except storage
I absolutlty love your solution for emulators, but add storage emulator please

**Comment by gpfister** [comment]
This is also a big requirement on our side, Storage plays a big role in our app. Please add Storage as soon as possible.

**Comment by LaMbA3** [comment]
+1

**Comment by mrudling** [comment]
Please add the storage emulator, really need this.

**Comment by therealmodbom** [comment]
a big +1 for me too

**Comment by pbilka46** [comment]
:+1: ! I would also like to have storage emulator

**Comment by abeisgoat** [comment]
Hey folks, if you're following this thread and willing to be an Alpha tester of "Bug fixes and Performance Improvements 😉" related to this bug, follow these steps:

1. Sign up for the [Firebase Alpha program](https://services.google.com/fb/forms/firebasealphaprogram/) 
2. Send an email to abehaskins@google.com, make sure to send it from the email address you will use for testing.

Please only do this if you're seriously interested in trying out an early-stage, buggy "performance improvement" and providing feedback! There will be issues and some parts of the setup will be hard, but we want to know what you think.

**Comment by eru123** [comment]
Is there any update about this feature request? badly needed this to test file upload features of my project.

**Comment by warting** [comment]
> Is there any update about this feature request? badly needed this to test file upload features of my project.

Looks like it is just around the corner :)
Check the commits included in v9.8.0

**Comment by oittaa** [comment]
`customTime` seems to be missing from stored metadata https://github.com/firebase/firebase-tools/blob/93abee929d79b2216a31af98871b6d29588810be/src/emulator/storage/metadata.ts

https://cloud.google.com/storage/docs/metadata#custom-time

For example here's how the Python client uses `customTime`:

https://github.com/googleapis/python-storage/blob/440a0a4ffe00b1f7c562b0e9c1e47dbadeca33e1/google/cloud/storage/blob.py#L102

**Comment by samtstern** [comment]
@warting follow the instructions that @abeisgoat posted here if you're interested:
https://github.com/firebase/firebase-tools/issues/1738#issuecomment-807520420

@oittaa let's not use this issue for feedback, there are too many people subscribed here and adding comments will create a lot of noise. You can contact @abeisgoat with feedback using the instructions above or if that is not working for some reason you can file a new issue on this repo.

**Comment by eduardolat** [comment]
+1 it's really important to us to have a COMPLETE emulator suite wrapping all the firebase services. We need to make local tests and integrate with CI/CD but if one is missing we can't build good tests around our entire application.

**Comment by gursesl** [comment]
Hoping for an announcement on [Google I/O](https://events.google.com/io/?lng=en) ;)

**Comment by mono0926** [comment]
I’ve found the Google I/O session🐶

Local development using the new Storage Emulator: https://events.google.com/io/session/21e0d8b3-cf63-4931-a6d7-ab8a8e2b10c9

**Comment by abeisgoat** [comment]
@mono0926 ssssh :wink:

**Comment by abeisgoat** [comment]
Hey folks, at Google I/O we'll be announcing that the Cloud Storage for Firebase emulator, released in `firebase-tools@9.11.0`, is available in beta today. Thank you for your patience and support. 

As someone who has worked on Firebase for nearly a decade it's thrilling to see the excitement of the community and to be able to keep delivering features that we only dreamed of back when we were a 10 person startup working from an apartment. We always wanted to offer a way for folks to build with Firebase entirely offline and with the Cloud Storage emulator we're finally filling in the last major gap for the Emulator Suite. It's been a long time coming! 

Hope you enjoy, excited to see all the bugs you report,
abe
